http/peer-id-auth.md

```diff
@@ -0,0 +1,331 @@
+| data to sign (hex encoded)                                                                  | 6c69627032702d5065657249443d6368616c6c656e67652d7365727665723d455245524552455245524552455245524552455245524552455245524552455245524552455245524552453d36636c69656e742d7075626c69632d6b65793d080112208139770ea87d175f56a35466c34c7ecccb8d8a91b4ee37a25df60f5b8fc9b39414686f73746e616d653d6578616d706c652e636f6d |
+| signature (base64 encoded)                                                                  | UA88qZbLUzmAxrD9KECbDCgSKAUBAvBHrOCF2X0uPLR1uUCF7qGfLPc7dw3Olo-LaFCDpk5sXN7TkLWPVvuXAA==                                                                                                                                                                                                                       |
+
+Note that the `=` after the libp2p-PeerID scheme is actually the varint length of the challenge-server parameter.
+
+## Base64 Encoding
+
+The base64 encoding follows Base 64 Encoding with URL and Filename Safe
+Alphabet from [RFC
+4648](https://datatracker.ietf.org/doc/html/rfc4648#section-5). Padding MAY be
+omitted. The reason this is not a multibase is to aid clients or servers who
+can not or prefer not to import a multibase dependency.
+
+## Public Key Encoding
+
+The authentication below exchanges the peer's public key instead of its PeerID,
+as the PeerID alone may not be enough to validate a signature. The Public Key is
+encoded per the [Peer ID spec] under the section "Keys" section.
+
+## Mutual Client and Server Peer ID Authentication
+
+The following protocol allows both the client and server to authenticate each
+other's Peer ID by having them each sign a challenge issued by the other. The
+protocol operates as follows:
+
+### Server Initiated Handshake
+
+1. The client makes an HTTP request to an authenticated resource.
+
+2. The server responds with status code 401 (Unauthorized) and set the header:
+   ```
+   WWW-Authenticate: libp2p-PeerID challenge-client="<challenge-string>", public-key="<base64-encoded-public-key-bytes>", opaque="<opaque-value>"
+   ```
+
+   The public-key parameter is the server's public key. It is the same public
+   key used to derive the server's peer id.
+
+   The opaque parameter is opaque to client. The client MUST return the opaque
+   parameter back to the server. The server MAY use the opaque parameter to
+   encode state.
+
+3. The client makes another HTTP request to the same authenticated resource and
+   sets the header:
+
+   ```
+   Authorization: libp2p-PeerID public-key="<base64-encoded-public-key-bytes>", opaque="<opaque-from-server>", challenge-server="<challenge-string>", sig="<base64-signature-bytes>"
```

Comment:
It's clear from context but do you want to rename `sig` to `sig-client` and `sig-server`?

Reply:
No preference. Happy to change, but I'd prefer calling the client-sig and server-sig.

Reply:
I am fine with keeping this as sig.